### ISSUE TITLE
imbl::HashMap is not covariant over the Key type.

### ISSUE THREAD

**Issue by maartendeprez** [jneem/imbl#53]
When using a type with a lifetime parameter as key for `imbl::HashMap`, read-only methods like get will not accept keys with shorter lifetimes. For example:

```rust
use imbl::HashMap;
use std::borrow::Cow;

#[derive(Clone, Debug, Hash, PartialEq, Eq)]
struct Key<'a> {
    key1: Cow<'a, str>,
    key2: Cow<'a, str>,
}

fn main() {
    let mut map: HashMap<Key<'static>, String> = HashMap::new();
    map.insert(
        Key {
            key1: Cow::Borrowed("some key 1"),
            key2: Cow::Borrowed("some key 2"),
        },
        String::from("some value"),
    );

    let key1 = String::from("some key 1");
    let key2 = String::from("some key 2");

    eprintln!(
        "{:?}",
        map.get(&Key {
            key1: Cow::Borrowed(&key1),
            key2: Cow::Borrowed(&key2)
        })
    );
}
```

When compiling this, the compiler complains that `key1` and `key2` do not live long enough; it requires that they are borrowed for 'static, even though it seems get should only need the key being available for the duration of the function call. I believe this is due to `imbl::HashMap` being invariant over its key parameter. The same example compiles and works as expected using `std::collections::HashMap`.

Would it be possible to make `imbl::HashMap` covariant over the key type? Or is there a reason this is not possible?

**Comment by jneem** [comment]
Thanks for the report. I'm pretty sure your diagnosis is correct, as shown by this smaller example which fails to compile (but compiles with `std::collections::HashMap`)
```rust
        fn assign<'a, 'b: 'a>(src: imbl::HashMap<&'b i32, i32>, mut dst: imbl::HashMap<&'a i32, i32>) {
            dst = src;
        }
```

I don't see any reason that `HashMap` cannot be covariant (over both types, right?), but I also had a quick look and didn't see any obvious place that the invariance is coming from. I will try to find some time tonight or tomorrow to dig into it more.

Presumably this is also an issue with the other collections in this crate.

**Comment by jneem** [comment]
Ok, I think I've figured out the reason for this: I tracked it down to the fact that `SparseChunk<A>` is invariant in `A`.

In more detail, we use the `sized_chunks` crate to manage various chunks of memory. `sized_chunks` uses type-level integers to express the sizes of these chunks, and various trait tricks to do computations on type-level integers. Unfortunately, traits in rust are always invariant. It may be possible to fix this by using const generics instead of type-level integers, but I guess I'll need to fork `sized_chunks` first.
